daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionCoderSpec.scala

```diff
@@ -605,6 +606,78 @@ class TransactionCoderSpec
       }
     }
 
+    "do Versioned" in {
+      forAll(Gen.oneOf(TransactionVersion.All), bytesGen, minSuccessful(5)) { (version, bytes) =>
+        val encoded = TransactionCoder.encodeVersioned(version, bytes)
+        val Right(decoded) = TransactionCoder.decodeVersioned(encoded)
+        decoded shouldBe Versioned(version, bytes)
+      }
+    }
+
+    "reject versioned message with trailing data" in {
+      forAll(
+        Gen.oneOf(TransactionVersion.All),
+        bytesGen,
+        bytesGen.filterNot(_.isEmpty),
+        minSuccessful(5),
+      ) { (version, bytes1, bytes2) =>
+        val encoded = TransactionCoder.encodeVersioned(version, bytes1)
+        TransactionCoder.decodeVersioned(encoded concat bytes2) shouldBe a[Left[_, _]]
+      }
+    }
+
+    "do FatContractInstance" in {
+      forAll(
+        malformedCreateNodeGen.filter(_.version >= TransactionVersion.V14),
+        timestampGen,
+        bytesGen,
+        minSuccessful(5),
+      ) { (create, time, salt) =>
+        val normalizedCreate = normalizeCreate(create)
+        val instance = FatContractInstance.fromCreateNode(
+          normalizedCreate,
+          time,
+          data.Bytes.fromByteString(salt),
+        )
+        val Right(encoded) = TransactionCoder.encodeFatContractInstance(instance)
+        val Right(decoded) = TransactionCoder.decodeFatContractInstance(encoded)
+
+        decoded shouldBe instance
+      }
+    }
+
+    "reject FatContractInstance with unknown fields" in {
+      forAll(
+        malformedCreateNodeGen.filter(_.version >= TransactionVersion.V14),
+        timestampGen,
+        bytesGen,
+        Arbitrary.arbInt.arbitrary,
+        bytesGen.filterNot(_.isEmpty),
+        minSuccessful(5),
+      ) { (create, time, salt, i, extraBytes) =>
+        val normalizedCreate = normalizeCreate(create)
+        val instance = FatContractInstance.fromCreateNode(
+          normalizedCreate,
+          time,
+          data.Bytes.fromByteString(salt),
+        )
+        val Right(encoded) = TransactionCoder.encodeFatContractInstance(instance)
+        val Right(Versioned(v, bytes)) = TransactionCoder.decodeVersioned(encoded)
+        val proto = TransactionOuterClass.FatContractInstance.parseFrom(bytes)
+        val x = addUnknownField(proto, i, extraBytes)
+        assert(proto != x)
```

Comment:
If the assert is part of the test, then that makes sense but then `x` could be reused below.

Reply:
Good catch.